File: sliding_window/sliding_window_max.py
from collections import deque
from typing import List


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        q = deque()
        l = r = 0
        while r < len(nums):
            # check if previous number is smaller, then pop it
            while q and nums[q[-1]] < nums[r]:
                q.pop()
            # insert the index in queue
            q.append(r)

            # check if left most val in queue is out of sliding wind bounds
            if l > q[0]:
                q.popleft()
            
            # if inside sliding wind, append max value and incr l
            if (r + 1) >= k:
                res.append(nums[q[0]])
                l = l + 1
            r = r + 1
        return res
    
print(Solution().maxSlidingWindow([7,2,4], 2))

File: sliding_window/sliding_window_max.js
// T - 0(n), M - O(n)
var maxSlidingWindow = function(nums, k) {
  const output = [];
  const q = [];
  let l = 0, r = 0;
  while(r < nums.length) {
      while(q && nums[q[q.length-1]] < nums[r]) {
          q.pop();
      }
      q.push(r);

      if(q[0] < l) {
          q.shift();
      }

      if(r+1-l == k) {
          output.push(nums[q[0]]);
          l += 1;
      }
      r += 1;
  }
  return output;
};